laplace/subnetlaplace.py

```diff
@@ -0,0 +1,129 @@
+        that define the subnetwork to apply the Laplace approximation over
+    sigma_noise : torch.Tensor or float, default=1
+        observation noise for the regression setting; must be 1 for classification
+    prior_precision : torch.Tensor or float, default=1
+        prior precision of a Gaussian prior (= weight decay);
+        can be scalar, per-layer, or diagonal in the most general case
+    prior_mean : torch.Tensor or float, default=0
+        prior mean of a Gaussian prior, useful for continual learning
+    temperature : float, default=1
+        temperature of the likelihood; lower temperature leads to more
+        concentrated posterior and vice versa.
+    backend : subclasses of `laplace.curvature.CurvatureInterface`
+        backend for access to curvature/Hessian approximations
+    backend_kwargs : dict, default=None
+        arguments passed to the backend on initialization, for example to
+        set the number of MC samples for stochastic approximations.
+    """
+    # key to map to correct subclass of BaseLaplace, (subset of weights, Hessian structure)
+    _key = ('subnetwork', 'full')
+
+    def __init__(self, model, likelihood, subnetwork_indices, sigma_noise=1., prior_precision=1.,
+                 prior_mean=0., temperature=1., backend=BackPackGGN, backend_kwargs=None):
+        self.H = None
+        super().__init__(model, likelihood, sigma_noise=sigma_noise, prior_precision=prior_precision,
+                         prior_mean=prior_mean, temperature=temperature, backend=backend,
+                         backend_kwargs=backend_kwargs)
+        # check validity of subnetwork indices and pass them to backend
+        self._check_subnetwork_indices(subnetwork_indices)
+        self.backend.subnetwork_indices = subnetwork_indices
+        self.n_params_subnet = len(subnetwork_indices)
+
+    def _init_H(self):
+        self.H = torch.zeros(self.n_params_subnet, self.n_params_subnet, device=self._device)
+
+    def _check_subnetwork_indices(self, subnetwork_indices):
+        """Check that subnetwork indices are valid indices of the vectorized model parameters
+           (i.e. `torch.nn.utils.parameters_to_vector(model.parameters())`).
+        """
+        if subnetwork_indices is None:
+            raise ValueError('Subnetwork indices cannot be None.')
+        elif not (isinstance(subnetwork_indices, torch.LongTensor) and\
+            subnetwork_indices.numel() > 0 and len(subnetwork_indices.shape) == 1):
+            raise ValueError('Subnetwork indices must be non-empty, 1-dimensional torch.LongTensor.')
+        elif not (len(subnetwork_indices[subnetwork_indices < 0]) == 0 and\
+            len(subnetwork_indices[subnetwork_indices >= self.n_params]) == 0):
+            raise ValueError(f'Subnetwork indices must lie between 0 and n_params={self.n_params}.')
+        elif not (subnetwork_indices.sort()[0].equal(torch.unique(subnetwork_indices, sorted=True))):
```

Comment:
`len(subnetwork_indices.unique()) == len(subnetwork_indices)` is probably a bit better to read and would do the same?

Reply:
Yes of course, good catch, changed it!